R/unit.R

```diff
@@ -145,258 +150,198 @@ calcStringMetric <- function(text) {
-
-stringUnit <- function(unit) {
-    unit %in% c("strwidth", "strheight", "strascent", "strdescent")
-}
-
-grobUnit <- function(unit) {
-    unit %in% c("grobx", "groby", "grobwidth", "grobheight",
-                "grobascent", "grobdescent")
-}
-
-dataUnit <- function(unit) {
-    stringUnit(unit) | grobUnit(unit)
-}
-
-recycle.data <- function(data, data.per, max.n, units) {
-    # FIRST STEP:  check that data needs to be recycled
-    if (any(dataUnit(units))) {
-        # VERY IMPORTANT:  Even if there is only one data specified
-        # and/or only one data needed, we want this to be a LIST of
-        # data values so that a single data and several data can be
-        # handled equivalently
-        # The test for whether it is only a single value currently
-        # consists of a check for mode="character" (i.e., a single
-        # string) or mode="expression" (i.e., a single expression)
-        # or class="grob" (i.e., a single grob) or class="gPath"
-        if (is.character(data) || is.language(data) ||
-            is.grob(data) || inherits(data, "gPath"))
-            data <- list(data)
-        if (data.per)
-            n <- max.n
-        else
-            n <- length(data)
-        original <- data
-        length(data) <- n
-        n.o <- length(original)
-        if (n.o < n)
-            for (i in (n.o + 1L):n)
-                data[[i]] <- original[[(i - 1L) %% n.o + 1L]]
-    }
-    data
-}
-
-# Make sure that and "str*" and "grob*" units have data
-valid.data <- function(units, data) {
-    n <- length(units)
-    str.units <- stringUnit(units)
-    if (any(str.units))
-        for (i in (1L:n)[str.units])
-            if (!(length(data) >= i &&
-                  (is.character(data[[i]]) || is.language(data[[i]]))))
-                stop("no string supplied for 'strwidth/height' unit")
-    # Make sure that a grob has been specified
-    grob.units <- grobUnit(units)
-    if (any(grob.units))
-        for (i in (1L:n)[grob.units]) {
-            if (!(length(data) >= i &&
-                  (is.grob(data[[i]]) || inherits(data[[i]], "gPath") ||
-                   is.character(data[[i]]))))
-                stop("no 'grob' supplied for 'grobwidth/height' unit")
-            if (is.character(data[[i]]))
-                data[[i]] <- gPath(data[[i]])
-            if (inherits(data[[i]], "gPath"))
-                if (depth(data[[i]]) > 1)
-                    stop("'gPath' must have depth 1 in 'grobwidth/height' units")
-        }
-    # Make sure that where no data is required, the data is NULL
-    if (!all(sapply(data[!(str.units | grob.units)], is.null)))
-        stop("non-NULL value supplied for plain unit")
-    data
-}
+                     "mylines", "mychar", "mystrwidth", "mystrheight",
+                     "sum", "min", "max")
 
 valid.units <- function(units) {
   .Call(C_validUnits, units)
 }
 
+# Printing, formating, and coercion
+unitDesc <- function(x, format = FALSE, ...) {
```

Comment:
So "unit arithmetic" will print out differently, right ?

Reply:
Yes, `sum`, `min`, and `max` units are printed in a special way where the data object is recursed through

Reply:
Cool.  Thanks